Lib/test/test_sqlite3/test_dbapi.py

```diff
@@ -583,6 +583,12 @@ def test_connect_positional_arguments(self):
             cx.close()
         self.assertEqual(cm.filename, __file__)
 
+    def test_connection_resource_warning(self):
+        with self.assertWarns(ResourceWarning):
+            cx = sqlite.connect(":memory:")
+            del cx
+            gc_collect()
```

Comment:
@vstinner: this test does not play well if run multiple times (for example during ref leak hunting); the `ResourceWarning` escapes the test during the second run, so we get an "env changed" warning and a failed test. I'm not sure how to fix this.

Reply:
I cannot reproduce the issue. Do you have logs where it occurred? A way to reproduce the issue? I tried ``./python -m test -m test_connection_resource_warning -v test_sqlite3 test_sqlite3`` to run the test twice in the same process. Using a debug build, it doesn't fail.

Reply:
Hm, there's something fishy here. It's not _this_ particular test that is emitting resource warnings, it's something else. Try:

```
$ ./python -m test -R : test_sqlite3 -m test_dbapi
```

Reply:
Ok, I found them:

    ./python -m test -R : test_sqlite3 -m ClosedCurTests

Reply:
This is because we close a cursor, not a connection:
```diff
diff --git a/Lib/test/test_sqlite3/test_dbapi.py b/Lib/test/test_sqlite3/test_dbapi.py
index d80ad7af32..e99333625b 100644
--- a/Lib/test/test_sqlite3/test_dbapi.py
+++ b/Lib/test/test_sqlite3/test_dbapi.py
@@ -1757,6 +1757,7 @@ def test_closed(self):
             with self.assertRaises(sqlite.ProgrammingError):
                 method = getattr(cur, method_name)
                 method(*params)
+        con.close()
 
 
 class SqliteOnConflictTests(unittest.TestCase):
```
I can prepare PR in 2 minutes.

Reply:
Check out #108360.